Created a class for an IR sensor.

Citysurvivor opened the pull request:
Class converts voltage from a short range IR sensor into distance. Returns fixed values when the distance exceeds or falls below predetermined values.

See ir sensor notes in the wiki for instructions, tips.

Issue #15

Cl7Y committed:
Created a class for the IR sensor.

Converts voltage from the short range sensor into distance. #15

nrgill28 commented on src/main/java/frc/robot/sensors/IrSensor.java:
```diff
@@ -0,0 +1,66 @@
+  
+    public double getDistance() {
+        //Calculate distance (cm) from mV:
+        double currentDistance = ((1/(((m_filter.calculate(m_irsensor.getValue()))/1000 +irXValue)/irSlopeValue))+irConstant);
```
This line is pretty hard to read, maybe break it up a bit?
Also, the median filter might not be a good idea unless the sensor is constantly getting values.
For example: if the sensor is polled 20 times at 10cm, and then a minute later is polled once at 30cm the previous 9 polls are being included in this calculation when they are far off from the actual value.

kevlam2706 replied:
We could do two things..

1 - remove the median filter if we think this class will only be used for "ad hoc" readings

2 - add a "takeMeasurement()" method to the class and tell users to put a call to that in the periodic() method of whatever subsystem instantiates one of these. The method would then update the median filter.

Citysurvivor replied:
Added a takeMeasurement() method to be called in a periodic method. Updates the median filter and returns the voltage, if anyone wishes to use the latter.

nrgill28 commented on src/main/java/frc/robot/sensors/IrSensor.java:
```diff
@@ -0,0 +1,66 @@
+        if (currentDistance < irMinDistance) {
+            currentDistance = 0; //set distance to a fixed value if it is below the min measuring range.
+            //Warning: Distances under 2cm will still produce erroneous values.
+        }
```
I think these two if statements can be replaced with `currentDistance = Math.min(Math.max(currentDistance, irMinDistance), irMaxDistance)`

(Effectively a clamp function)

Citysurvivor replied:
Actually, I think I will keep the if statement. If the measured distance falls below the minimum value (4.5), I will set it to a different constant (0) instead of setting it to match the minimum value (4.5). I'm not sure if that can be done with that math.min/math.max code, so I'll stick with an if statement.

nrgill28 commented on src/main/java/frc/robot/sensors/IrSensor.java:
```diff
@@ -0,0 +1,66 @@
+    //median filter to discard outliers. Filters over 10 samples
+    private final MedianFilter m_filter = new MedianFilter(10);
+
+    private final AnalogInput m_irsensor;
```
This one is a little nitpicky since it's about naming conventions, but I don't see the use for the `m_` prefix. I'd just remove that.

kevlam2706 replied:
The m_ prefix is meant to indicate a member variable and was more popular as a formatting convention in my C++ days.

Yeah, if nobody else is using that convention then you can remove it.

Citysurvivor replied:
I think I'll keep that naming convention for now since I'm more used to it.

nrgill28 commented on src/main/java/frc/robot/sensors/IrSensor.java:
```diff
@@ -0,0 +1,66 @@
+    private final double irXValue = -0.007545;
+    private final double irConstant = -3.92;
+    private final double irMaxDistance = 35;
+    private final double irMinDistance = 4.5;
```
Are any of these tune-able? And if so, would it be beneficial to be able to edit them over NetworkTables without re-deploying the code? If so you may want to look at changing these into FluidConstants.

kevlam2706 replied:
No, these are meant to curve-fit based on the physical response of the sensor. They shouldn't change.

One suggestion: maybe include the units? e.g. irMaxDistanceCm

kevlam2706 replied:
Charlie can you write somewhere in the comments the exact model number of the sensor this class is meant to represent? Sharp model ABC123-whatever. Do you still have the data sheet on your computer? Otherwise I'll help you find that number.

Citysurvivor replied:
Variables won't change often (only if we swap out IR sensors), so there wouldn't be much benefit from changing those into FluidConstants.

Added sensor model number, and added units to variable names.

nrgill28 commented on src/main/java/frc/robot/sensors/IrSensor.java:
```diff
@@ -0,0 +1,66 @@
+
+public class IrSensor implements Sendable {
+    private static final int kIRPort = 1; //set IR sensor port
+        //TODO: The IR sensor's port may change, depending on the robot setup. (Use a robot-specific config to set this?)
```
Yes, the channel for the sensor should be defined in the Config.java file with a robotSpecific value. This way it can automatically switch between channels based on which robot it's deployed to

kevlam2706 replied:
I think this should be a value passed into the constructor so I can do this:

```
IrSensor inputSensor = new IrSensor(0); // input sensor is connected to analog input 0
IrSensor outputSensor = new IrSensor(2); // output sensor is connected to analog input 2;
```

Citysurvivor replied:
Yup, fixed that. Channel number can now be passed as an argument.

nrgill28 reviewed (commented):
Great work! Code looks pretty good overall, just a few comments here and there on a couple things

kevlam2706 reviewed (approved):
Looks good!

georgetzavelas commented on src/main/java/frc/robot/sensors/IrSensor.java:
```diff
@@ -0,0 +1,71 @@
+    //Set the formula values to calculate IR sensor distance:
+        //Values for short range (Sharp GP2Y0A41SK0F) 4-30cm sensor: 12.8528, -0.007545, -3.92, 30, 4.5.
+        //Note: If a different IR sensor is used, these values will need to be updated.
+    private final double irSlopeValue = 12.8528;
```
Since these are constants do you mind CAPITALIZING them?

Citysurvivor replied:
Sure thing. Also, I believe words in constants are separated by underscores, i.e. IR_SLOPE_VALUE, so I'll do that.

georgetzavelas commented on src/main/java/frc/robot/sensors/IrSensor.java:
```diff
@@ -0,0 +1,71 @@
+
+        //The following "if" statements fix the values outside the sensor's accurate measuring range.
+        if (currentDistanceCm > irMaxDistanceCm) {
+            currentDistanceCm = 30; //set distance to a fixed value if it is above the max measuring range.
```
Can we make this number to the const you defined above?

Citysurvivor replied:
> Can we make this number to the const you defined above?

OK, did that for max distance. 

As for minimum distance, I think the code should set currentDistanceCm to 0 if the measured value falls below the minimum distance. The reasoning being that, if an object was inside the blindspot between the sensor and the sensor's minimum range, it would be safer to set the variable to zero (and assume the object is touching) than to assume the object is located around the minimum range.

georgetzavelas commented on src/main/java/frc/robot/sensors/IrSensor.java:
```diff
@@ -0,0 +1,71 @@
+            currentDistanceCm = 30; //set distance to a fixed value if it is above the max measuring range.
+        }
+        if (currentDistanceCm < irMinDistanceCm) {
+            currentDistanceCm = 0; //set distance to a fixed value if it is below the min measuring range.
```
Can you define a new const for this? It will make the code more readable.

Citysurvivor replied:
@georgetzavelas Not sure if I understand. Should I use a variable in place of the zero? I think currentDistanceCm should always be set to zero if the measured distance is below the sensor's minimum range.

Citysurvivor replied:
OK, used a constant in place of zero.

georgetzavelas commented on src/main/java/frc/robot/sensors/IrSensor.java:
```diff
@@ -0,0 +1,71 @@
+    }
+
+    public IrSensor(int channel) {
+        this.m_irsensor = new AnalogInput(kIRPort);
```
Channel is being passed in here but I don't see it being used.

Citysurvivor replied:
Whoops, fixed that. Channel will be used to set the IR sensor channel instead of having it hardcoded.

georgetzavelas reviewed (commented):
Great job Charlie! Some comments for your consideration.

Cl7Y committed:
Improved readability, comment documentation, and fixed a bug.

Cl7Y committed:
Variable tweaks and comment updates.

Cl7Y committed:
Added a takeMeasurement method to get continuous measurements in periodic methods.

Cl7Y committed:
Minor edits.

Cl7Y committed:
Added a takeMeasurement method. Un-hardcoded sensor port variable. Tweaked variables.

Added a takeMeasurement method, which can be placed in a periodic method to continuously update the median filter and the IR sensor's voltage outputs.

IR sensor port is no longer hard coded. Channel is passed as an argument into the class' constructor.

Tweaked variables, variable names, and comments, for accuracy.

georgetzavelas commented on src/main/java/frc/robot/sensors/IrSensor.java:
```diff
@@ -0,0 +1,80 @@
+/* the project.                                                               */
+/*----------------------------------------------------------------------------*/
+
+/**
```
Nice touch but this is not needed since git has a version history of who modified what and when. Also it is misleading since someone else could come and change a few lines in this file.

Citysurvivor replied:
OK, will keep in mind for future commits. Won't change it for now since it would require the merge request to be re-approved, and the issue is fairly minor .

georgetzavelas reviewed (approved):
Great refactor!
